processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java

```diff
@@ -495,15 +501,49 @@ else if ( Executables.isFieldAccessor( candidate ) && ( Executables.isFinal( can
             }
 
             Accessor previousCandidate = result.get( targetPropertyName );
-            if ( previousCandidate == null || preferredType == null || ( targetType != null
-                && typeUtils.isAssignable( preferredType.getTypeMirror(), targetType.getTypeMirror() ) ) ) {
+
+
+            if (previousCandidate != null
+                    && sourceAccessor != null
+                    && sourceAccessor.get( targetPropertyName ) != null
+                    && isBetter( previousCandidate, candidate,
+                    sourceAccessor.get( targetPropertyName ).getAccessedType() ) ) {
+                result.put( targetPropertyName, candidate );
+                continue;
+            }
+
+
+            if (previousCandidate == null || preferredType == null || (targetType != null
+                    && typeUtils.isAssignable( preferredType.getTypeMirror(), targetType.getTypeMirror() ) ) ) {
                 result.put( targetPropertyName, candidate );
             }
         }
 
         return result;
     }
 
+    private TypeMirror extractTypeMirror( Accessor accessor ) {
+        return accessor.getExecutable().getParameters().get( 0 ).asType();
```

Comment:
In theory you should not be able to get up to here if your `Accessor` is field. However, I think that a `null` check needs to be done on the `accessor.getExecutable()`.